File: src/java/BinarySearch.java
package com.attamusc.practice;

import java.util.ArrayList;
import java.util.Arrays;

class BinarySearch {
  public static int search(int query, ArrayList<Integer> a) {
    int mid = a.size() / 2;
    int start = 0;
    int end = a.size() - 1;

    while (start <= end) {
      mid = start + (end - start) / 2;
      System.out.println(mid);

      if (query > a.get(mid)) {
        start = mid + 1;
      }
      else if (query < a.get(mid)) {
        end = mid - 1;
      }
      else {
        return mid;
      }
    }
    return -1;
  }

  public static void main(String[] args) {
    int index = BinarySearch.search(5, new ArrayList<Integer>(Arrays.asList(1, 2, 2, 3, 4, 5, 6)));
    System.out.println("Answer: " + index);
  }
}

File: src/scala/binary_search.scala
package com.attamusc.practice

object BinarySearch {
  def search(query: Int, a: List[Int]) : Int = {
    var mid : Int = a.size / 2
    var start = 0
    var end = a.size - 1

    while (start <= end) {
      mid = start + (end - start) / 2
      println(mid)
      a(mid) match {
        case i if query > i => start = mid + 1
        case i if query < i => end = mid - 1
        case i if query == i => return mid
      }
    }
    return -1
  }
}

// val index = BinarySearch.search(5, List(1, 2, 2, 3, 4, 5, 6))
// println("Answer: " + index)
